### PR TITLE
kv: Replace multiTestContext with TestCluster in consistency_queue_test

### PR DESCRIPTION
**Pull request by lunevalex**
Makes progress on #8299

multiTestContext is legacy construct that is deprecated in favor of running tests via TestCluster. This is one PR out of many to remove the usage of multiTestContext in the consistency_queue test cases.

Release note : None

### CODE REVIEW

**Review by tbg — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50257)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @lunevalex, and @tbg)

___
*[pkg/kv/kvserver/consistency_queue_test.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50257#-MA0cOYhDSDffh4ikcvC:-MA0cOYiAp0KaYHCYx2i:bi1s10o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3aaaa794e6049da1db57d04f393b30014b03749/pkg/kv/kvserver/consistency_queue_test.go#L73)):*
> ```Go
> 	// observe a timestamp in the past to trick the clock.
> 	ts.Clock().Update(hlc.Timestamp{
> 		WallTime: ts.Clock().PhysicalNow() + store.LeaseExpiration(),
> ```

Thanks for working on this!
We should not be passing in bogus future timestamps into clocks, not even in tests. The reason is that our system is build on assumptions of bounded clock skew, which is certainly violated here. The test would likely survive this (it may panic sporadically due when the clock skew is detected) but as a general rule we want to avoid this pattern, at least in its naive form. Being able to expire leases is a necessary tool to get the test to run faster in some cases, though. When we do need it, we could look into running a test cluster with a very large MaxOffset (larger than the lease expiration), which would allow us to bump this at least once. I'm sure there's a can of worms there somewhere, though.

This particular test would benefit from a simpler refactor: make the consistency checker's `shouldQueue` method unit testable and don't start any nodes at all.

Basically, take https://github.com/cockroachdb/cockroach/blob/76c63ded01f18aa31c08ed6a2e9cf4972b0b0459/pkg/kv/kvserver/consistency_queue.go#L77-L79 and replace `*Replica` with a `replica interface { getQueueLastProcessed() (hlc.Timestamp, error); Desc() *roachpb.RangeDescriptor) }`, pass a similarly taylored interface in for `NodeLiveness` and when all is done you get to test the desired functionality directly. For inspiration, check out `raft_log_queue.go` and `replica_gc_queue.go`, both of them do this sort of thing.

**Review by lunevalex — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50257)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[pkg/kv/kvserver/consistency_queue_test.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50257#-MA0cOYhDSDffh4ikcvC:-MA1YgSrDP1KR4sgyWsV:b-4dfhti) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3aaaa794e6049da1db57d04f393b30014b03749/pkg/kv/kvserver/consistency_queue_test.go#L73)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Thanks for working on this!
We should not be passing in bogus future timestamps into clocks, not even in tests. The reason is that our system is build on assumptions of bounded clock skew, which is certainly violated here. The test would likely survive this (it may panic sporadically due when the clock skew is detected) but as a general rule we want to avoid this pattern, at least in its naive form. Being able to expire leases is a necessary tool to get the test to run faster in some cases, though. When we do need it, we could look into running a test cluster with a very large MaxOffset (larger than the lease expiration), which would allow us to bump this at least once. I'm sure there's a can of worms there somewhere, though.

This particular test would benefit from a simpler refactor: make the consistency checker's `shouldQueue` method unit testable and don't start any nodes at all.

Basically, take https://github.com/cockroachdb/cockroach/blob/76c63ded01f18aa31c08ed6a2e9cf4972b0b0459/pkg/kv/kvserver/consistency_queue.go#L77-L79 and replace `*Replica` with a `replica interface { getQueueLastProcessed() (hlc.Timestamp, error); Desc() *roachpb.RangeDescriptor) }`, pass a similarly taylored interface in for `NodeLiveness` and when all is done you get to test the desired functionality directly. For inspiration, check out `raft_log_queue.go` and `replica_gc_queue.go`, both of them do this sort of thing.
</blockquote></details>

Thank you for the suggestion this makes sense, I will refactor and update.

**Comment by lunevalex** [comment]
___
*[pkg/kv/kvserver/consistency_queue_test.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50257#-MA0cOYhDSDffh4ikcvC:-MAT1pnt5WIv7UT3O-yu:b-rrdzuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3aaaa794e6049da1db57d04f393b30014b03749/pkg/kv/kvserver/consistency_queue_test.go#L73)):*
<details><summary><i>Previously, lunevalex wrote…</i></summary><blockquote>

Thank you for the suggestion this makes sense, I will refactor and update. 
</blockquote></details>

It ended up being more messy than originally intended because of kvserver vs kvserver_test, but the test is a lot simpler now.

File: pkg/kv/kvserver/consistency_queue.go [hunk 1]
```diff
@@ -77,30 +86,50 @@ func newConsistencyQueue(store *Store, gossip *gossip.Gossip) *consistencyQueue
 func (q *consistencyQueue) shouldQueue(
 	ctx context.Context, now hlc.Timestamp, repl *Replica, _ *config.SystemConfig,
 ) (bool, float64) {
-	interval := q.interval()
-	if interval <= 0 {
+	return consistencyQueueShouldQueueImpl(ctx, now,
```
Comment (tbg) [inline pkg/kv/kvserver/consistency_queue.go:89]: Curious why `now` isn't also on `consistencyShouldQueueData`.

File: pkg/kv/kvserver/consistency_queue_test.go [hunk 2]
```diff
@@ -234,46 +241,64 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
...
-+    value:"\x00\x00\x00\x00\x01T"
-+    raw mvcc_key/value: 6500000000000000007b000003db0d 000000000154
-`
-			if act != exp {
-				// We already logged the actual one above.
-				t.Errorf("expected:\n%s\ngot:\n%s", exp, act)
-			}
-
+			// With a real clock its not possible to check the actual diff,
+			// so we settle for just a key check.
```
Comment (tbg) [inline pkg/kv/kvserver/consistency_queue_test.go:245]: Why can't you add `diff[0].Timestamp.WallTime=123` to the old code?

File: pkg/kv/kvserver/helpers_test.go [hunk 3]
```diff
@@ -91,10 +90,18 @@ func (s *Store) ComputeMVCCStats() (enginepb.MVCCStats, error) {
 
 // ConsistencyQueueShouldQueue invokes the shouldQueue method on the
 // store's consistency queue.
-func (s *Store) ConsistencyQueueShouldQueue(
-	ctx context.Context, now hlc.Timestamp, r *Replica, cfg *config.SystemConfig,
+func ConsistencyQueueShouldQueue(
```
Comment (tbg) [inline pkg/kv/kvserver/helpers_test.go:93]: why is this exported?

Comment (tbg) [reply pkg/kv/kvserver/helpers_test.go:93]: (and why does it even exist? On master, it's called only from `TestConsistencyQueueRequiresLive` which you've refactored).

**Review by tbg — APPROVED** [review]
LGTM with some comments, thanks!

File: pkg/kv/kvserver/helpers_test.go (hunk 3 above)
Comment (lunevalex) [reply pkg/kv/kvserver/helpers_test.go:93]: TestConsistencyQueueRequiresLive stills calls this and it lives in kvserver_test. Unfortunately we cant move that file into kvserver because of the circular dependency between TestCluster and kvserver.

File: pkg/kv/kvserver/consistency_queue.go (hunk 1 above)
Comment (lunevalex) [reply pkg/kv/kvserver/consistency_queue.go:89]: My mindeset was to wrap everything that we use from Replica in this struct, and now() was part of the original method signature so I left it out. No strong preference on this one, could go either way.

File: pkg/kv/kvserver/consistency_queue_test.go (hunk 2 above)
Comment (lunevalex) [reply pkg/kv/kvserver/consistency_queue_test.go:245]: My understanding is that the timestamp is part of the raw MVCC record, which makes this fairly hard to diff. Here is an example with 2 consecutive runs of the test with this code restored to as it. 

`    consistency_queue_test.go:257: expected:
        --- leaseholder
        +++ follower
        +0.000000123,987 "e"
        +    ts:1970-01-01 00:00:00.000000123 +0000 UTC
        +    value:"\x00\x00\x00\x00\x01T"
        +    raw mvcc_key/value: 6500000000000000007b000003db0d 000000000154
        
        got:
        --- leaseholder
        +++ follower
        +1593109825.736774000,987 "e"
        +    ts:2020-06-25 18:30:25.736774 +0000 UTC
        +    value:"\x00\x00\x00\x00\x01T"
        +    raw mvcc_key/value: 6500161bdcf223399170000003db0d 000000000154


            consistency_queue_test.go:257: expected:
        --- leaseholder
        +++ follower
        +0.000000123,987 "e"
        +    ts:1970-01-01 00:00:00.000000123 +0000 UTC
        +    value:"\x00\x00\x00\x00\x01T"
        +    raw mvcc_key/value: 6500000000000000007b000003db0d 000000000154
        
        got:
        --- leaseholder
        +++ follower
        +1593109975.875626000,987 "e"
        +    ts:2020-06-25 18:32:55.875626 +0000 UTC
        +    value:"\x00\x00\x00\x00\x01T"
        +    raw mvcc_key/value: 6500161bdd151832a410000003db0d 000000000154`

Comment (lunevalex) [reply pkg/kv/kvserver/consistency_queue_test.go:245]: Specifically note: `raw mvcc_key/value: 6500161bdcf223399170000003db0d 000000000154` vs `raw mvcc_key/value: 6500161bdd151832a410000003db0d 000000000154`

Comment (tbg) [reply pkg/kv/kvserver/consistency_queue_test.go:245]: Sorry, this dropped off my radar. I would "just" regexp-replace `s~(raw mvcc_key/value: ).*~\1 <omitted>~g` so that we compare the parts we know should match.

File: pkg/kv/kvserver/helpers_test.go (hunk 3 above)
Comment (tbg) [reply pkg/kv/kvserver/helpers_test.go:93]: Makes sense!

File: pkg/kv/kvserver/consistency_queue_test.go (hunk 2 above)
Comment (lunevalex) [reply pkg/kv/kvserver/consistency_queue_test.go:245]: I think in that case the only thing that is comparable is the value, since everything else has a timestamp, so we can just compare those directly without using strings.

**Comment by tbg** [comment]
It's also a crude way to test the output format. I know it's not ideal, if
we test this more directly elsewhere then yeah definitely don't bother
salvaging this

File: pkg/kv/kvserver/consistency_queue_test.go (hunk 2 above)
Comment (lunevalex) [reply pkg/kv/kvserver/consistency_queue_test.go:245]: hmm, I just overrode the timestamp and that seemed to have solved it all. Ok going with that.
